### PR TITLE
Source-map support for multiple input source files

### PR DESCRIPTION
**Pull request by divmain**
This PR allows tools built on top of Babel to parse multiple JS sources to AST, optionally transform them,  and combine them.  When output code is generated, source-maps will point to the original source files.

The minimum change necessary for [my use case](https://github.com/interlockjs/interlock) is contained in the first commit.  However, the additional changes add support for this use case to both ends of the Babel toolchain (and removes unnecessary traversals of the AST outside of Babel).

Documentation was added where it seemed it would be helpful.  I did not add tests since the test helpers assume a 1:1 transformation.  If you'd like me to add tests, let me know what it should look like.

Thanks!

### CODE REVIEW

**Comment by amasad** [comment]
Why not concat the files after the transform as opposed to before transform? And then combine the sourcemaps. I know that using the combine-source-map module is awfully slow but you can use this neat trick:

https://github.com/facebook/react-native/blob/master/packager/react-packager/src/Bundler/Bundle.js#L199-L232

It didn't have full browser support last I checked, but let me know if that solves it for you

**Comment by divmain** [comment]
Hey @amasad, thanks for the link.  That's an interesting work-around.  I'm not sure it will work for me - I'm building a bundler similar in function to webpack, and losing broad browser support would be a significant drawback.

The change proposed here is really the only thing keeping Babel from being used in place of something like `recast`.  And it is also something that `escodegen` supports - I've been using that until recently, but would really prefer to use Babel for everything.  When we discussed awhile back in the Slack channel, it seemed like there was an openness to the change.

There's definitely risk in opening Babel to new use-cases like this.  And I also recognize that there may not be a lot in it for the Babel maintainers :)  But if there's anything I can do to make it a more attractive change, I'm definitely open.  Since it is a minimal and non-breaking change, I hope that there's a possibility.

File: packages/babylon/src/parser/node.js [hunk 1]
```diff
@@ -6,11 +6,12 @@ import { SourceLocation } from "../util/location";
 const pp = Parser.prototype;
 
 class Node {
-  constructor(pos?: number, loc?: SourceLocation) {
+  constructor(pos?: number, loc?: SourceLocation, filename?: any) {
```
Comment (amasad) [inline packages/babylon/src/parser/node.js]: `filename?: string`

File: packages/babylon/src/parser/node.js [hunk 2]
```diff
@@ -6,11 +6,12 @@ import { SourceLocation } from "../util/location";
...
     this.type = "";
     this.start = pos;
     this.end = 0;
     this.loc = new SourceLocation(loc);
+    if (filename) { this.loc.filename = filename; }
```
Comment (amasad) [inline packages/babylon/src/parser/node.js]: no need for the curlys

**Comment by amasad** [comment]
OK. Sounds good. Can you please add tests?

**Pull request merged**